File: ci/release.py
"""CLI to help prepare and publish release.

To prepare stable release:

```bash
python ci/release.py version patch  # or 'minor' or 'major'
```

To prepare dev release:

```bash
python ci/release.py version patch --dev
```

To publish stable release:

```bash
python ci/release.py publish
```

To publish dev release:

```bash
python ci/release.py publish --dev
```

"""

import json
from argparse import ArgumentParser
from pathlib import Path
from subprocess import run
from textwrap import dedent


def run_cli(cmd, *args, **kwargs):
    cmd_joined = " ".join(cmd)
    print(f"> {cmd_joined}")
    return run(cmd, *args, **kwargs, check=True)


def check_unreleased_has_items(changelog_content: str):
    """Check that there are items in the Unreleased section."""

    items_in_unreleased = []
    lines = changelog_content.splitlines()
    idx = 0
    while idx < len(lines):
        if lines[idx] != "## Unreleased":
            idx += 1
            continue
        # Find lines under unreleased
        idx += 1
        while idx < len(lines):
            if lines[idx].startswith("##"):
                break
            if lines[idx] and lines[idx].startswith("-"):
                items_in_unreleased.append(lines[idx])
            idx += 1

    for item in items_in_unreleased:
        if "No unreleased changes" in item:
            raise RuntimeError("Please update 'No unreleated changes' with changelog items.")

    if not items_in_unreleased:
        raise RuntimeError("Please add changelog items under the 'Unreleased' header.")


def check_git_clean():
    """Check that git status is clean."""
    git_status = run_cli(["git", "status", "--porcelain"], text=True, capture_output=True)
    if git_status.stdout != "":
        raise RuntimeError(f"git status is not clean:\n{git_status.stdout}")


def get_current_js_version():
    package_path = Path("modal-js") / "package.json"
    with package_path.open("r") as f:
        json_package = json.load(f)
        return json_package["version"]


def update_version(args):
    """Updates version and changelog and prepare a release PR."""
    if args.update not in ["major", "minor", "patch"]:
        raise RuntimeError("update parameter must be 'major', 'minor', or 'patch'")

    check_git_clean()

    if args.dev:
        current_version = get_current_js_version()

        if "-dev." in current_version:
            run_cli(["npm", "version", "prerelease", "--no-git-tag-version"], cwd="modal-js")
        else:
            run_cli(["npm", "version", f"pre{args.update}", "--preid=dev", "--no-git-tag-version"], cwd="modal-js")

        new_version = get_current_js_version()

        run_cli(["git", "diff"])

        commit_message = f"[DEV-RELEASE] Prepare dev release for modal-js/v{new_version}"
        if args.dry_run:
            print("\nDRY RUN: Would create commit with message:")
            print(commit_message)
            run_cli(["git", "restore", "--", "modal-js/package.json"])
        else:
            run_cli(["git", "add", "modal-js/package.json"])
            run_cli(["git", "commit", "-m", commit_message])
    else:
        changelog_path = Path("CHANGELOG.md")
        changelog_content = changelog_path.read_text()
        check_unreleased_has_items(changelog_content)

        run_cli(["npm", "version", args.update, "--no-git-tag-version"], text=True, cwd="modal-js")
        new_version = get_current_js_version()

        version_header = f"modal-js/v{new_version}"

        new_header = dedent(f"""\
        ## Unreleased

        No unreleased changes.

        ## {version_header}""")

        new_changelog_content = changelog_content.replace("## Unreleased", new_header)
        changelog_path.write_text(new_changelog_content)

        run_cli(["git", "diff"])
        run_cli(["git", "add", "modal-js/package.json", str(changelog_path)])

        commit_message = f"[RELEASE] Prepare release for {version_header}"
        if args.dry_run:
            print("\nDRY RUN: Would create commit with message:")
            print(commit_message)
            run_cli(["git", "reset", "HEAD"])
            run_cli(
                ["git", "restore", "--", "modal-js/package.json", str(changelog_path)]
            )
        else:
            run_cli(["git", "commit", "-m", commit_message])


def publish(args):
    """Publish modal-js"""
    version = get_current_js_version()
    git_tags = [f"{version}", f"modal-js/v{version}"]

    if args.dry_run:
        print("\nDRY RUN: Would execute the following operations:")
        print("- git push (push version commit)")
        print(f"- Create and push git tags: {' '.join(git_tags)}")
        if args.dev:
            print("- npm publish --tag next (in modal-js/)")
        else:
            print("- npm publish (in modal-js/)")
        return

    run_cli(["git", "push"])
    for tag in git_tags:
        run_cli(["git", "tag", tag])
    run_cli(["git", "push", "--tags"])

    if args.dev:
        run_cli(["npm", "publish", "--tag", "next"], cwd="modal-js")
    else:
        run_cli(["npm", "publish"], cwd="modal-js")



def main():
    """Entrypoint for preparing and publishing release."""
    parser = ArgumentParser()
    subparsers = parser.add_subparsers(required=True)

    version_parser = subparsers.add_parser("version")
    version_parser.add_argument("update")
    version_parser.add_argument("--dev", action="store_true", help="Create dev release")
    version_parser.add_argument("--dry-run", action="store_true", help="Show what would be done without making changes")
    version_parser.set_defaults(func=update_version)

    publish_parser = subparsers.add_parser("publish")
    publish_parser.add_argument("--dev", action="store_true", help="Publish dev release")
    publish_parser.add_argument("--dry-run", action="store_true", help="Show what would be done without making changes")
    publish_parser.set_defaults(func=publish)

    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: scripts/release.ts
/**
 * @description リリース準備とパブリッシュのCLI
 *
 * Usage:
 *   bun scripts/release.ts version patch          # stable release
 *   bun scripts/release.ts version patch --dev    # dev release
 *   bun scripts/release.ts publish                # publish stable
 *   bun scripts/release.ts publish --dev          # publish dev
 *   Add --dry-run to preview without making changes.
 */

import { execSync } from "node:child_process";
import { readFileSync, writeFileSync } from "node:fs";
import { parseArgs } from "node:util";

function run(cmd: string) {
	console.log(`> ${cmd}`);
	return execSync(cmd, { stdio: "inherit" });
}

function runCapture(cmd: string): string {
	console.log(`> ${cmd}`);
	return execSync(cmd, { encoding: "utf-8" }).trim();
}

function getVersion(): string {
	const pkg = JSON.parse(readFileSync("package.json", "utf-8"));
	return pkg.version;
}

function checkGitClean() {
	const status = runCapture("git status --porcelain");
	if (status !== "") {
		throw new Error(`git status is not clean:\n${status}`);
	}
}

function checkUnreleasedHasItems(content: string) {
	const lines = content.split("\n");
	let inUnreleased = false;
	const items: string[] = [];

	for (const line of lines) {
		if (line === "## Unreleased") {
			inUnreleased = true;
			continue;
		}
		if (inUnreleased && line.startsWith("##")) break;
		if (inUnreleased && line.startsWith("-")) items.push(line);
	}

	for (const item of items) {
		if (item.includes("No unreleased changes")) {
			throw new Error(
				"Please update 'No unreleased changes' with changelog items.",
			);
		}
	}
	if (items.length === 0) {
		throw new Error(
			"Please add changelog items under the 'Unreleased' header.",
		);
	}
}

function version(update: string, dev: boolean, dryRun: boolean) {
	if (!["major", "minor", "patch"].includes(update)) {
		throw new Error("update must be 'major', 'minor', or 'patch'");
	}

	checkGitClean();

	if (dev) {
		const current = getVersion();
		if (current.includes("-dev.")) {
			run("npm version prerelease --no-git-tag-version");
		} else {
			run(`npm version pre${update} --preid=dev --no-git-tag-version`);
		}

		const newVersion = getVersion();
		run("git diff");

		const msg = `[DEV-RELEASE] Prepare dev release for v${newVersion}`;
		if (dryRun) {
			console.log(`\nDRY RUN: Would create commit with message:\n${msg}`);
			run("git restore -- package.json");
		} else {
			run("git add package.json");
			run(`git commit -m "${msg}"`);
		}
	} else {
		const changelog = readFileSync("CHANGELOG.md", "utf-8");
		checkUnreleasedHasItems(changelog);

		run(`npm version ${update} --no-git-tag-version`);
		const newVersion = getVersion();
		const header = `v${newVersion}`;

		const newChangelog = changelog.replace(
			"## Unreleased",
			`## Unreleased\n\nNo unreleased changes.\n\n## ${header}`,
		);
		writeFileSync("CHANGELOG.md", newChangelog);

		run("git diff");
		run("git add package.json CHANGELOG.md");

		const msg = `[RELEASE] Prepare release for ${header}`;
		if (dryRun) {
			console.log(`\nDRY RUN: Would create commit with message:\n${msg}`);
			run("git reset HEAD");
			run("git restore -- package.json CHANGELOG.md");
		} else {
			run(`git commit -m "${msg}"`);
		}
	}
}

function publish(dev: boolean, dryRun: boolean) {
	const ver = getVersion();
	const tags = [ver, `v${ver}`];

	if (dryRun) {
		console.log("\nDRY RUN: Would execute the following operations:");
		console.log("- git push (push version commit)");
		console.log(`- Create and push git tags: ${tags.join(" ")}`);
		console.log(dev ? "- npm publish --tag next" : "- npm publish");
		return;
	}

	run("git push");
	for (const tag of tags) {
		run(`git tag ${tag}`);
	}
	run("git push --tags");

	if (dev) {
		run("npm publish --tag next");
	} else {
		run("npm publish");
	}
}

const { positionals, values } = parseArgs({
	args: process.argv.slice(2),
	allowPositionals: true,
	options: {
		dev: { type: "boolean", default: false },
		"dry-run": { type: "boolean", default: false },
	},
});

const command = positionals[0];
const isDev = values.dev ?? false;
const isDryRun = values["dry-run"] ?? false;

switch (command) {
	case "version": {
		const update = positionals[1];
		if (!update)
			throw new Error(
				"Usage: bun scripts/release.ts version <patch|minor|major>",
			);
		version(update, isDev, isDryRun);
		break;
	}
	case "publish":
		publish(isDev, isDryRun);
		break;
	default:
		throw new Error("Usage: bun scripts/release.ts <version|publish>");
}
